### ISSUE TITLE
During AugLagrangian's optimization, L_BFGS sets coordinates to nan

### ISSUE THREAD

**Issue by olgavrou** [mlpack/ensmallen#365]
#### Issue description

During `AugLagrangian`'s optimization, `L_BFGS` sets coordinates to `nan`

I believe the issue is [here](https://github.com/mlpack/ensmallen/blob/master/include/ensmallen_bits/lbfgs/lbfgs_impl.hpp#L89)

```
scalingFactor = dot(sMat, yMat) / dot(yMat, yMat);
```

if the two dot products are very small this leads to the scaling factor becoming a `nan`

In my experiment, if I replace that line with the code pasted below:

```
scalingFactor = dot(sMat, yMat) / std::max(1e-12, dot(yMat, yMat));
```

(EDITED)

or

```
double max_y = arma::abs(yMat).max();

if (max_y == 0.0)
{
  scalingFactor = 0.0;
}
else
{
  auto z = yMat / max_y;
  scalingFactor = (dot(sMat, z) / max_y) / dot(z, z);
}
```

which results in the optimizer working as expected.

Please let me know if you think this calculation is ok for the library and if you would be open to either patching it or accepting a patch.

#### Your environment

 * version of ensmallen: 2.19.1
 * Ubuntu 20.04:
 * compiler: gcc 9.4.0
 * version of Armadillo: 12.1.90
 * any other environment information you think is relevant:

#### Steps to reproduce

I have been iterating over the implementation of my constraint optimization problem you can see the latest version [here](https://github.com/VowpalWabbit/vowpal_wabbit/pull/4581/files) and the behaviour is seen when [this](https://github.com/VowpalWabbit/vowpal_wabbit/pull/4581/files#diff-77e8284a33c61d33a5de316722e95a09a60773df3574240f7a21dd6e90a34cf8L593) test (`CheckUpdateRule500WIterations`) is [run](https://github.com/VowpalWabbit/vowpal_wabbit/actions/runs/4942865717/jobs/8836792141) 

but unfortunately I don't have a smaller repro in hand

#### Expected behavior

Coordinates not getting set to `nan`

#### Actual behavior

Coordinates get set to `nan`

**Comment by olgavrou** [comment]
OK will do :) I think the first solution is probably working better, I got another `nan` with the second one at some point

**Comment by conradsnicta** [comment]
> `scalingFactor = dot(sMat, yMat) / std::max(1e-12, dot(yMat, yMat));`

Instead of a hardcoded value like 1e-12, suggest to use someting like  
`1000 * std::numeric_limits<CubeElemType>::epsilon()`.  

The user may elect to use matrices and cubes with single-precision floating point values, rather than double-precision.  In other words, `CubeElemType` can be either `float` or `double`.  For single-precision, the hardcoded 1e-12 is probably too low in this context.

**Comment by conradsnicta** [comment]
PS. While we are at it, the following line should be replaced:
https://github.com/mlpack/ensmallen/blob/master/include/ensmallen_bits/lbfgs/lbfgs_impl.hpp#LL93C1-L93C57

     scalingFactor = 1.0 / sqrt(dot(gradient, gradient));

with

     scalingFactor = 1.0 / arma::norm(gradient, "fro");

This is for two reasons: (1) clarity of intent, (2) Armadillo will use a more robust algorithm to calculate the norm.

**Comment by conradsnicta** [comment]
I've proposed workarounds in #368

**Comment by conradsnicta** [comment]
Should be resolved in #368. If not, please re-open and provide more details.
